optuna/samplers/_nsga2/crossover.py

```diff
@@ -0,0 +1,503 @@
+    dominates: Callable[[FrozenTrial, FrozenTrial, Sequence[StudyDirection]], bool],
+) -> List[FrozenTrial]:
+    m = select_m(crossover_name)
+    parents = []
+    for _ in range(m):
+        parent = _select_parent(
+            study, [t for t in parent_population if t not in parents], rng, dominates
+        )
+        parents.append(parent)
+    return parents
+
+
+def _select_parent(
+    study: Study,
+    parent_population: Sequence[FrozenTrial],
+    rng: np.random.RandomState,
+    dominates: Callable[[FrozenTrial, FrozenTrial, Sequence[StudyDirection]], bool],
+) -> FrozenTrial:
+    # TODO(ohta): Consider to allow users to specify the number of parent candidates.
+    population_size = len(parent_population)
+    candidate0 = parent_population[rng.choice(population_size)]
+    candidate1 = parent_population[rng.choice(population_size)]
+
+    # TODO(ohta): Consider crowding distance.
+    if dominates(candidate0, candidate1, study.directions):
+        return candidate0
+    else:
+        return candidate1
+
+
+def _uniform(
+    x1: np.ndarray, x2: np.ndarray, rng: np.random.RandomState, swapping_prob: float
+) -> np.ndarray:
+    # https://www.researchgate.net/publication/201976488_Uniform_Crossover_in_Genetic_Algorithms
+    # Section 1 Introduction
+
+    assert x1.shape == x2.shape
+    assert len(x1.shape) == 1
+
+    child = []
+    for x1_i, x2_i in zip(x1, x2):
+        param = x1_i if rng.rand() < swapping_prob else x2_i
+        child.append(param)
+    return np.array(child)
+
+
+def _blxalpha(
+    x1: np.ndarray, x2: np.ndarray, rng: np.random.RandomState, alpha: float
+) -> np.ndarray:
+    # http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.465.6900&rep=rep1&type=pdf
+    # Section 2 Crossover Operators for RCGA 2.1 Blend Crossover
+
+    assert x1.shape == x2.shape
+    assert len(x1.shape) == 1
+
+    xs = np.stack([x1, x2])
+
+    x_min = xs.min(axis=0)
+    x_max = xs.max(axis=0)
+    diff = alpha * (x_max - x_min)  # Equation (1)
+    low = x_min - diff  # Equation (1)
+    high = x_max + diff  # Equation (1)
+    r = rng.uniform(0, 1, size=len(diff))
+    param = (high - low) * r + low
+    return param
+
+
+def _sbx(
+    x1: np.ndarray,
+    x2: np.ndarray,
+    rng: np.random.RandomState,
+    distributions: List[BaseDistribution],
+    eta: float,
+) -> np.ndarray:
+    # https://www.researchgate.net/profile/M-M-Raghuwanshi/publication/267198495_Simulated_Binary_Crossover_with_Lognormal_Distribution/links/5576c78408ae7536375205d7/Simulated-Binary-Crossover-with-Lognormal-Distribution.pdf
+    # Section 2 Simulated Binary Crossover (SBX)
+
+    # To avoid generating solutions that violate the box constraints,
+    # alpha1, alpha2, xls and xus is introduced, unlike the reference.
+    _xls = []
+    _xus = []
+    for distribution in distributions:
+        assert isinstance(distribution, _NUMERICAL_DISTRIBUTIONS)
+        _xls.append(distribution.low)
+        _xus.append(distribution.high)
```

Comment:
Ditto about prefixed underscores. Could you either remove them or use different names if variable shadowing is an issue?